state/state.go

```diff
@@ -34,19 +42,128 @@ type DownloadedObject struct {
+}
+
+func NewDynamoDBStater(session *session.Session, service string, backfill int) *DynamoDBStater {
+
+	return &DynamoDBStater{
+		Session:          session,
+		Service:          service,
+		BackfillInterval: time.Hour * time.Duration(backfill),
+	}
+}
+
+// Used for unmarshaling and adding objects to DynamoDB
+type Record struct {
+	S3Object string
+	Time     time.Time
+	TTL      int64
+}
+
+// list of processed objects
+func (d *DynamoDBStater) ProcessedObjects() (map[string]time.Time, error) {
+	objs := make(map[string]time.Time)
+
+	var records []Record
+
+	svc := dynamodb.New(d.Session)
+	err := svc.ScanPages(&dynamodb.ScanInput{
+		TableName: aws.String(DynamoTableName),
+	}, func(logs *dynamodb.ScanOutput, last bool) bool {
+		recs := []Record{}
+
+		err := dynamodbattribute.UnmarshalListOfMaps(logs.Items, &recs)
+
+		// break out of function
+		if err != nil {
+			logrus.WithField("error", err).Debug("Failed to unmarshal DynamoDB Scan Items")
+			return false
+		}
+
+		records = append(records, recs...)
+		for _, rec := range records {
+			// break out of the scan, we've reached the end of our
+			// backfill interval
+			if time.Since(rec.Time) > d.BackfillInterval {
+				return false
+			}
+		}
+
+		return true
+	})
+
+	if err != nil {
+		return objs, fmt.Errorf("Error scanning DynamoDB, %v", err)
+	}
+
+	for _, record := range records {
+		objs[record.S3Object] = record.Time
+	}
+
+	return objs, nil
+}
+
+func (d *DynamoDBStater) SetProcessed(s3object string) error {
+
+	// add access log information
+	svc := dynamodb.New(d.Session)
+
+	objMap := Record{
+		S3Object: s3object,
+		Time:     time.Now(),
+		TTL:      time.Now().Add(TTLDefault).Unix(), // make this 7 days
+	}
+
+	obj, err := dynamodbattribute.MarshalMap(objMap)
+
+	if err != nil {
+		return fmt.Errorf("Marshalling DynamoDB object failed: %s", err)
+	}
+
+	// add object to dynamodb using conditional
+	// if the object exists, no write happens
+	input := &dynamodb.PutItemInput{
+		Item:                obj,
+		TableName:           aws.String(DynamoTableName),
+		ConditionExpression: aws.String("attribute_not_exists(S3Object)"),
+	}
+
+	_, err = svc.PutItem(input)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			// we want this to happen if object already exists
+			if awsErr.Code() != dynamodb.ErrCodeConditionalCheckFailedException {
+				return fmt.Errorf("PutItem failed: %s", err)
+			}
+
+		}
+		// if it is the conditional check, we can just pop out and
+		// ignore this object!
+		return nil
+	}
+
+	return nil
+}
+
 // FileStater is an implementation for indicating processing state using the
 // local filesystem for backing storage.
 type FileStater struct {
 	*sync.Mutex
-	StateDir string
-	Service  string
+	StateDir         string
+	Service          string
+	BackfillInterval time.Duration
 }
 
-func NewFileStater(stateDir, service string) *FileStater {
+func NewFileStater(stateDir, service string, backfill int) *FileStater {
```

Comment:
same suggestion here as before - `backfillHrs int` to make it more clear what's being held in the variable